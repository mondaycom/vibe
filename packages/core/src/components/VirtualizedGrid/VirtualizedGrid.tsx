import React, { CSSProperties, forwardRef, useCallback, useEffect, useMemo, useRef, useState } from "react";
import cx from "classnames";
import {
  GridChildComponentProps,
  GridOnScrollProps,
  ScrollDirection,
  VariableSizeGrid as Grid,
  GridOnItemsRenderedProps
} from "react-window";
import AutoSizer from "react-virtualized-auto-sizer";
import {
  getNormalizedItems,
  getOnItemsRenderedData,
  isLayoutDirectionScrollbarVisible
} from "../../services/virtualized-service";
import usePrevious from "../../hooks/usePrevious";
import useThrottledCallback from "../../hooks/useThrottledCallback";
import useMergeRef from "../../hooks/useMergeRef";
import VibeComponentProps from "src/types/VibeComponentProps";
import { VibeComponent } from "../../types";
import { NOOP } from "../../utils/function-utils";
import { ComponentDefaultTestId, getTestId } from "../../tests/test-ids-utils";
import styles from "./VirtualizedGrid.module.scss";
import { VirtualizedGridItemType as ItemType } from "./VirtualizedGrid.types";

export interface VirtualizedGridProps extends VibeComponentProps {
  /**
   * The list of items to be rendered in the grid.
   */
  items: ItemType[];
  /**
   * Function that renders each item in the grid.
   */
  itemRenderer: (item: ItemType, index: number, style: CSSProperties) => ItemType | GridChildComponentProps<ItemType>;
  /**
   * Function that returns the row height.
   */
  getRowHeight: () => number;
  /**
   * Function that returns the column width.
   */
  getColumnWidth: () => number;
  /**
   * Function that returns the unique ID of an item.
   */
  getItemId?: (item: ItemType, index: number) => string;
  /**
   * The index of the item to scroll to.
   */
  scrollToId?: number;
  /**
   * Callback fired when the grid is scrolled.
   */
  onScroll?: (horizontalScrollDirection: ScrollDirection, scrollTop: number, scrollUpdateWasRequested: boolean) => void;
  /**
   * Callback fired when scrolling has finished.
   */
  onScrollToFinished?: () => void;
  /**
   * Callback fired when items are rendered in the grid.
   */
  onItemsRendered?: ({
    firstItemId,
    secondItemId,
    lastItemId,
    centerItemId,
    firstItemOffsetEnd,
    currentOffsetTop
  }: {
    firstItemId: string;
    secondItemId: string;
    lastItemId: string;
    centerItemId: string;
    firstItemOffsetEnd: number;
    currentOffsetTop: number;
  }) => void;
  /**
   * The delay (in ms) for throttling the `onItemsRendered` callback.
   */
  onItemsRenderedThrottleMs?: number;
  /**
   * Callback fired when the grid size is updated.
   */
  onSizeUpdate?: (width: number, height: number) => void;
  /**
   * Callback fired when the vertical scrollbar visibility changes.
   */
  onVerticalScrollbarVisiblityChange?: (value: boolean) => void;
  /**
   * Class name applied to the scrollable container.
   */
  scrollableClassName?: string;
}

const VirtualizedGrid: VibeComponent<VirtualizedGridProps> = forwardRef(
  (
    {
      className,
      id,
      items = [],
      itemRenderer = (item: ItemType, _index: number, _style: CSSProperties) => item,
      getRowHeight = () => 50,
      getColumnWidth = () => 100,
      getItemId = (item: ItemType, _index: number) => item.id,
      onScroll,
      scrollToId = null,
      onScrollToFinished = NOOP,
      onItemsRendered = null,
      onItemsRenderedThrottleMs = 200,
      onSizeUpdate = NOOP,
      onVerticalScrollbarVisiblityChange = null,
      scrollableClassName,
      "data-testid": dataTestId
    }: VirtualizedGridProps,
    ref
  ) => {
    // states
    const [gridHeight, setGridHeight] = useState(0);
    const [gridWidth, setGridWidth] = useState(0);

    // prevs
    const prevScrollToId = usePrevious(scrollToId);

    // Refs
    const componentRef = useRef(null);
    const isVerticalScrollbarVisibleRef = useRef(null);
    const gridRef = useRef(null);
    const scrollTopRef = useRef(0);
    const animationDataRef = useRef({
      scrollOffsetInitial: 0,
      scrollOffsetFinal: 0,
      animationStartTime: 0
    });
    const mergedRef = useMergeRef(ref, componentRef);

    const animationData = animationDataRef.current;

    // Callbacks
    const heightGetter = useCallback(
      (item: ItemType) => {
        const height = getRowHeight();
        if (!height || Number.isNaN(height)) {
          console.error("Couldn't get height for item: ", item);
        }
        return height;
      },
      [getRowHeight]
    );

    const idGetter = useCallback(
      (item: ItemType, index: number) => {
        const itemId = getItemId(item, index);
        if (itemId === undefined) {
          console.error("Couldn't get id for item: ", item);
        }
        return itemId;
      },
      [getItemId]
    );

    // Memos
    // Creates object of itemId => { item, index, height, offsetTop}
    const normalizedItems = useMemo(() => {
      return getNormalizedItems(items, idGetter, heightGetter);
    }, [items, idGetter, heightGetter]);

    const calcColumnCount = useMemo(() => {
      return Math.min(items.length, Math.floor(gridWidth / getColumnWidth()));
    }, [items, gridWidth, getColumnWidth]);

    const calcRowCount = useMemo(() => {
      return calcColumnCount > 0 ? Math.ceil(items.length / calcColumnCount) : 0;
    }, [items, calcColumnCount]);

    const scrollToColumnIndex = useMemo(() => {
      return scrollToId % calcColumnCount;
    }, [scrollToId, calcColumnCount]);

    const scrollToRowIndex = useMemo(() => {
      return Math.floor(scrollToId / calcColumnCount);
    }, [scrollToId, calcColumnCount]);

    // Callbacks
    const onScrollCB = useCallback(
      ({ horizontalScrollDirection, scrollTop, scrollUpdateWasRequested }: GridOnScrollProps) => {
        scrollTopRef.current = scrollTop;
        if (!scrollUpdateWasRequested) {
          animationData.scrollOffsetInitial = scrollTop;
        }
        onScroll && onScroll(horizontalScrollDirection, scrollTop, scrollUpdateWasRequested);
      },
      [onScroll, scrollTopRef, animationData]
    );

    const cellRenderer = useCallback(
      ({ columnIndex, rowIndex, style }: { columnIndex: number; rowIndex: number; style: CSSProperties }) => {
        const index = rowIndex * calcColumnCount + columnIndex;
        const item = items[index];
        return itemRenderer(item, index, style);
      },
      [items, itemRenderer, calcColumnCount]
    );

    const updateGridSize = useCallback(
      (width: number, height: number) => {
        if (height !== gridHeight || width !== gridWidth) {
          setTimeout(() => {
            setGridHeight(height);
            setGridWidth(width);
            onSizeUpdate(width, height);
          }, 0);
        }
      },
      [gridHeight, gridWidth, onSizeUpdate]
    );

    const onItemsRenderedCB = useThrottledCallback(
      ({ visibleRowStartIndex, visibleRowStopIndex }: GridOnItemsRenderedProps) => {
        if (!onItemsRendered) return;
        const data = getOnItemsRenderedData(
          items,
          normalizedItems,
          idGetter,
          visibleRowStartIndex,
          visibleRowStopIndex,
          gridHeight,
          scrollTopRef.current
        );
        onItemsRendered(data);
      },
      { wait: onItemsRenderedThrottleMs, trailing: true },
      [onItemsRendered, items, normalizedItems, idGetter, gridHeight]
    );

    // Effects
    useEffect(() => {
      // scroll to specific item
      if (scrollToId && prevScrollToId !== scrollToId) {
        gridRef.current.scrollToItem({
          align: "center",
          columnIndex: scrollToColumnIndex,
          rowIndex: scrollToRowIndex
        });
        onScrollToFinished();
      }
    }, [scrollToId, prevScrollToId, gridRef, scrollToColumnIndex, scrollToRowIndex, onScrollToFinished]);

    useEffect(() => {
      // recalculate row heights
      if (gridRef.current) {
        gridRef.current.resetAfterIndices({ columnIndex: 0, rowIndex: 0 });
      }
    }, [normalizedItems]);

    useEffect(() => {
      // update vertical scrollbar visibility
      if (onVerticalScrollbarVisiblityChange) {
        const isVisible = isLayoutDirectionScrollbarVisible(items, normalizedItems, idGetter, gridHeight);
        if (isVerticalScrollbarVisibleRef.current !== isVisible) {
          isVerticalScrollbarVisibleRef.current = isVisible;
          onVerticalScrollbarVisiblityChange(isVisible);
        }
      }
    }, [onVerticalScrollbarVisiblityChange, items, normalizedItems, gridHeight, idGetter]);

    return (
      <div
        ref={mergedRef}
        className={cx(styles.virtualizedGridWrapper, className)}
        id={id}
        data-testid={dataTestId || getTestId(ComponentDefaultTestId.VIRTUALIZED_GRID, id)}
      >
        <AutoSizer>
          {({ height, width }: { height: number; width: number }) => {
            updateGridSize(width, height);
            return (
              <Grid
                ref={gridRef}
                height={height}
                width={width}
                columnWidth={getColumnWidth}
                columnCount={calcColumnCount}
                rowHeight={getRowHeight}
                rowCount={calcRowCount}
                onScroll={onScrollCB}
                onItemsRendered={onItemsRenderedCB}
                className={scrollableClassName}
              >
                {cellRenderer as unknown as VibeComponent<GridChildComponentProps>}
              </Grid>
            );
          }}
        </AutoSizer>
      </div>
    );
  }
);

export default VirtualizedGrid;
