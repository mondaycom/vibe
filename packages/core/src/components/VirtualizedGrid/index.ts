export { default as VirtualizedGrid, type VirtualizedGridProps } from "./VirtualizedGrid";
export * from "./VirtualizedGrid.types";
