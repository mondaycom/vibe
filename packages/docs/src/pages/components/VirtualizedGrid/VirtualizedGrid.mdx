import { Canvas, Meta } from "@storybook/blocks";
import { UsageGuidelines } from "vibe-storybook-components";
import * as VirtualizedGridStories from "./VirtualizedGrid.stories";

<Meta of={VirtualizedGridStories} />

# VirtualizedGrid

VirtualizedGrid is a component which only renders visible grid items, it is a logic component and doesn't change and look and feel

Under the hood we are using - [react-window](https://github.com/bvaughn/react-window) and [react-virtualized-auto-sizer](https://github.com/bvaughn/react-virtualized-auto-sizer)

<Canvas of={VirtualizedGridStories.Overview} />

### Import

```js
import { VirtualizedGrid } from "@vibe/core";
```

## Props

<PropsTable />

## Usage

<UsageGuidelines
  guidelines={[
    "Use this component to implement a grid with many items or when each item render has heavy calculations.",
    "Rendering only the visible grid items instead of all the grid's items will create better performance and a smoother experience for users while using the grid."
  ]}
/>

## Related components

<RelatedComponents componentsNames={["Table", "VirtualizedList", "MenuGridItem"]} />
