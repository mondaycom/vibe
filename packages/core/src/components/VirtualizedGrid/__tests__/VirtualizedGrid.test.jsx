import { it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import VirtualizedGrid from "../VirtualizedGrid";

it("renders correctly with empty props", () => {
  const tree = renderer.create(<VirtualizedGrid />).toJSON();
  expect(tree).toMatchSnapshot();
});
it("renders correctly with class name", () => {
  const tree = renderer.create(<VirtualizedGrid className="class-name" />).toJSON();
  expect(tree).toMatchSnapshot();
});
it("renders correctly with scrollable class name", () => {
  const tree = renderer.create(<VirtualizedGrid scrollableClassName="class-name" />).toJSON();
  expect(tree).toMatchSnapshot();
});
