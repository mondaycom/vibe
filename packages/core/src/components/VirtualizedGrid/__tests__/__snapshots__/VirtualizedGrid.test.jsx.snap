// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly with class name 1`] = `
<div
  className="virtualizedGridWrapper class-name"
  data-testid="virtualized-grid"
>
  <div
    style={
      {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  />
</div>
`;

exports[`renders correctly with empty props 1`] = `
<div
  className="virtualizedGridWrapper"
  data-testid="virtualized-grid"
>
  <div
    style={
      {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  />
</div>
`;

exports[`renders correctly with scrollable class name 1`] = `
<div
  className="virtualizedGridWrapper"
  data-testid="virtualized-grid"
>
  <div
    style={
      {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  />
</div>
`;
